### PR TITLE
(6/n) Support 2D Parallelism - Trainer example

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds an example for 2D parallelism (Tensor Parallelism + FSDP2) for the Trainer. It is equivalent to the one for Fabric added in #19846, so the model code etc. is copy-pasted. The main file to review is `examples/pytorch/tensor_parallel/train.py`.

This PR depends on #19878 

----
📚 Documentation preview 📚: https://pytorch-lightning--19879.org.readthedocs.build/en/19879/

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: examples/pytorch/tensor_parallel/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+Initializing distributed: GLOBAL_RANK: 2, MEMBER: 3/4
+----------------------------------------------------------------------------------------------------
+distributed_backend=nccl
+All distributed processes registered. Starting with 4 processes
+----------------------------------------------------------------------------------------------------
+
+LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [4,5,6,7]
+LOCAL_RANK: 3 - CUDA_VISIBLE_DEVICES: [4,5,6,7]
+LOCAL_RANK: 1 - CUDA_VISIBLE_DEVICES: [4,5,6,7]
+LOCAL_RANK: 2 - CUDA_VISIBLE_DEVICES: [4,5,6,7]
```
Comment (justusschock) [inline examples/pytorch/tensor_parallel/README.md:39]: why is it using 4,5,6,7 here or is that just because you used those to run it?

**Review by justusschock — APPROVED** [review]

File: examples/pytorch/tensor_parallel/model.py
Comment (lantiga) [inline examples/pytorch/tensor_parallel/model.py:1]: The license that is mentioned here is unfortunately not compatible with the Apache 2.0 license.

However looking here:
https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama/model.py

it is distributed under the BSD license
```
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree.
#
# Llama 2 is licensed under the LLAMA 2 Community License,
# Copyright (c) Meta Platforms, Inc. All Rights Reserved.
```

Can we include that file instead? It looks very close.

File: examples/pytorch/tensor_parallel/README.md (hunk 1 above)
Comment (awaelchli) [reply examples/pytorch/tensor_parallel/README.md]: yes, because I share the machine and run on these GPUs. I modified it now to show 0,1,3,4 not to be confusing.

File: examples/pytorch/tensor_parallel/model.py
Comment (awaelchli) [reply examples/pytorch/tensor_parallel/model.py:1]: Sure, I switched it out with this version now. It's the same apart from trivial aesthetic differences, so I don't know why the license can suddenly be different.
To clarify, the reason why I took Llama2 was because I was specifically asked to do so.

Comment (lantiga) [reply examples/pytorch/tensor_parallel/model.py:1]: Great thank you! Yeah it may be a nitpick, but I’d rather be 100% with licenses.

**Review by lantiga — APPROVED** [review]
🚀

File: examples/pytorch/tensor_parallel/train.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        parallelize(self.model, device_mesh=self.device_mesh)
+
+    def on_train_start(self) -> None:
+        self.model.init_weights()
+
+    def training_step(self, batch):
+        inputs = batch[:, :-1]
+        labels = batch[:, 1:]
+        output = self.model(inputs)
+        with loss_parallel():
```
Comment (carmocca) [inline examples/pytorch/tensor_parallel/train.py:30]: TIL. Are you also enabling this for backward? https://github.com/pytorch/pytorch/blob/5fb11cda4fe60c1a7b30e6c844f84ce8933ef953/torch/distributed/tensor/parallel/loss.py#L35
